Realm/RLMQueryUtil.mm

```diff
@@ -446,41 +498,42 @@ void add_numeric_constraint(RLMPropertyType datatype,
     template <typename A, typename B>
     void add_bool_constraint(RLMPropertyType, NSPredicateOperatorType operatorType, A&& lhs, B&& rhs);
 
-    void add_substring_constraint(null, Query condition);
     template<typename T>
     void add_substring_constraint(const T& value, Query condition);
     template<typename T>
     void add_substring_constraint(const Columns<T>& value, Query condition);
 
-    template <typename T>
+    template <typename C, typename T>
     void add_string_constraint(NSPredicateOperatorType operatorType,
                                NSComparisonPredicateOptions predicateOptions,
-                               Columns<String> &&column,
+                               Columns<C>&& column,
                                T value);
 
+    template <typename C, typename T>
+    void add_diacritic_sensitive_string_constraint(NSPredicateOperatorType operatorType,
+                                                   NSComparisonPredicateOptions predicateOptions,
+                                                   Columns<C>&& column,
+                                                   T value);
+
     template <typename R>
-    void add_constraint(RLMPropertyType type,
-                        NSPredicateOperatorType operatorType,
+    void add_constraint(NSPredicateOperatorType operatorType,
                         NSComparisonPredicateOptions predicateOptions,
                         ColumnReference const& column, R const& rhs);
-    template <typename... T>
+    template <template<typename> typename W, typename T>
```

Comment:
Can we give these template types better names? I can see how `W` is being used in the method body, but it seems like a fairly arbitrary designation that could use a bit more self documenting.

Reply:
W was initially Wrapper but that name wasn't very accurate and I failed to come up with a good name for it.